extension/test/background_test.ts

```diff
@@ -65,26 +87,76 @@ describe('background.ts', function () {
         { config: rcxMain.config }
       );
     });
+
+    it('should throw typeError if tab is undefined', async function () {
+      // }).to.throw(TypeError, 'sender.tab is always defined here.');
+    });
   });
-});
 
-async function sendMessageToBackground({
-  tabId = 0,
-  type,
-  responseCallback = () => {
-    // Do nothing by default.
-  },
-}: {
-  tabId?: number;
-  type: string;
-  responseCallback?: (response: unknown) => void;
-}): Promise<void> {
-  // In background.ts, a promise is passed to `addListener` so we can await it here.
-  // eslint-disable-next-line @typescript-eslint/await-thenable
-  await chrome.runtime.onMessage.addListener.yield(
-    { type: type },
-    { tab: { id: tabId } },
-    responseCallback
-  );
-  return;
-}
+  describe('xsearch', function () {
+    it('should return search results', async function () {
+      const searchStub = sinon.stub().returns(['to eat', 'verb']);
+      rcxMain.search = searchStub;
+      const request = {
+        type: 'xsearch',
+        text: 'た',
+        dictOption: '2',
+      };
+      const sender = { tab: { id: 0 } };
+      const response = sinon.spy();
+
+      // eslint-disable-next-line @typescript-eslint/await-thenable
+      await chrome.runtime.onMessage.addListener.callArgWith(
+        0,
+        request,
+        sender,
+        response
+      );
+
+      expect(searchStub.calledWith(request.text, request.dictOption)).to.be
```

Comment:
I wrote the other comment for information but in generally verify intermediate state isn't necessary if you are testing the inputs and outputs.

In this case, you can drop this and you can know it's working by just testing the response is called with the right arguments.

If you want to ensure the right arguments are passed you can update the stub rather than using a verification. Usually, though keeping stubs loose is good for behavior testing. 

(this is pretty subtle so it's something to think about each test and build up some intuition)

Reply:
So in this case, I would have to test that the response callback  is being called with a class method that passes in the arguments?`rcxMain.search(request.text, request.dictOption)` ?

Can we just change the test to: `expect(searchStub.calledWith(request.text, sinon.match.any)).to.be.true;`

and then drop:
`expect(response.called).to.be.true` and call it a day?

Reply:
> So in this case, I would have to test that the response callback is being called with a class method that passes in the arguments? `rcxMain.search(request.text, request.dictOption)` ?

This doesn't sound right. `response` is called with the return value of `RcxMain.search` which you have stubbed at the top of the test. So you just need to confirm it's called with the values you passed to `sinon.stub`

Or I misunderstood! I don't think any class methods are involved...

> and call it a day?

Actually, the opposite is usually what we want from a readability/maintainability point of view.

drop `expect(searchStub.calledWith...)`

And add: `expect(response).toBeCalledWith(...)`

the amount of code and structure is basically the same but the points of focus are the different:

With the way I'm suggesting, you're passing through the inputs and confirming the outputs. This is contrast to verifying some internal method call which is hidden from the user.